### PR TITLE
Improvement: Refactoring of the transport module

### PR DESCRIPTION
**Pull request by petardimitrijevic**
After more detailed analysis of the transport module it looks to me that there is no need for separate structures for different type of transport. Instead everything can be done with a common CoapEndpoint structure with different constructors.

I've refactored the code to remove the transport module and put everything into a single file.
It simplifies the implementation and removes lot of boiler plate code. In addition instead of having multiple unsafe functions, there is only one function call which is unsafe now.

I've looked into future TLS implementation as well, and this change won't affect that, when the time for implementation comes.

This PR changes the internal architecture of the library.  I'm not certain if this acceptable for you.
Please feel free to reject this PR if it is not.

### CODE REVIEW

File: libcoap/src/context.rs [hunk 1]
```diff
@@ -232,7 +233,7 @@ impl CoapContext<'_> {
     pub fn add_endpoint_udp(&mut self, addr: SocketAddr) -> Result<(), EndpointCreationError> {
         // SAFETY: Because we never return an owned reference to the endpoint, it cannot outlive the
         // context it is bound to (i.e. this one).
-        let endpoint = unsafe { CoapUdpEndpoint::new(self, addr)? }.into();
+        let endpoint = CoapEndpoint::new_udp_endpoint(self, addr)?.into();
```
Comment (JKRhb) [inline libcoap/src/context.rs:236]:
```suggestion
        let endpoint = CoapEndpoint::new_udp_endpoint(self, addr)?;
```

File: libcoap/src/context.rs [hunk 2]
```diff
@@ -232,7 +233,7 @@ impl CoapContext<'_> {
     pub fn add_endpoint_udp(&mut self, addr: SocketAddr) -> Result<(), EndpointCreationError> {
         // SAFETY: Because we never return an owned reference to the endpoint, it cannot outlive the
         // context it is bound to (i.e. this one).
```
Comment (JKRhb) [inline libcoap/src/context.rs:235]: Is this comment still needed without the `unsafe` block?

File: libcoap/src/context.rs [hunk 3]
```diff
@@ -259,7 +260,7 @@ impl CoapContext<'_> {
     pub fn add_endpoint_dtls(&mut self, addr: SocketAddr) -> Result<(), EndpointCreationError> {
         // SAFETY: Provided context (i.e., this instance) will not outlive the endpoint, as we will
         // drop it alongside this context and never hand out copies of it.
-        let endpoint = unsafe { CoapDtlsEndpoint::new(self, addr)? }.into();
+        let endpoint = CoapEndpoint::new_dtls_endpoint(self, addr)?.into();
```
Comment (JKRhb) [inline libcoap/src/context.rs:263]:
```suggestion
        let endpoint = CoapEndpoint::new_dtls_endpoint(self, addr)?;
```

File: libcoap/src/context.rs [hunk 4]
```diff
@@ -244,7 +245,7 @@ impl CoapContext<'_> {
     pub fn add_endpoint_tcp(&mut self, addr: SocketAddr) -> Result<(), EndpointCreationError> {
         // SAFETY: Because we never return an owned reference to the endpoint, it cannot outlive the
         // context it is bound to (i.e. this one).
```
Comment (JKRhb) [inline libcoap/src/context.rs:247]: Same here: Still required?

File: libcoap/src/context.rs [hunk 5]
```diff
@@ -259,7 +260,7 @@ impl CoapContext<'_> {
     pub fn add_endpoint_dtls(&mut self, addr: SocketAddr) -> Result<(), EndpointCreationError> {
         // SAFETY: Provided context (i.e., this instance) will not outlive the endpoint, as we will
         // drop it alongside this context and never hand out copies of it.
```
Comment (JKRhb) [inline libcoap/src/context.rs:262]: Same here

File: libcoap/src/transport.rs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+    /// Method utilized by transport protocol specific constructors to actually create the endpoint in libcoap
+    fn new_endpoint(
+        context: &mut CoapContext,
+        addr: SocketAddr,
+        proto: coap_proto_t,
+    ) -> Result<Self, EndpointCreationError> {
+        let endpoint = unsafe {
+            // SAFETY: coap_new_endpoint will return null if it is unable to add new endpoint.
+            // These states are processed further in the code
```
Comment (JKRhb) [inline libcoap/src/transport.rs:45]: Should this comment be inside or before the unsafe block?

**Review by JKRhb — COMMENTED** [review]
Hi @petardimitrijevic, thank you for this PR! :) The changes look good to me, but as I am not a maintainer of this crate, other project members need to make a decision here (CC @pulsastrix, @falko17). See some initial comments/suggestions below already, though. (As I am not an expert when it comes to unsafe Rust, I hope that they make sense.)

File: libcoap/src/context.rs [hunk 7]
```diff
@@ -244,7 +245,7 @@ impl CoapContext<'_> {
     pub fn add_endpoint_tcp(&mut self, addr: SocketAddr) -> Result<(), EndpointCreationError> {
         // SAFETY: Because we never return an owned reference to the endpoint, it cannot outlive the
         // context it is bound to (i.e. this one).
-        let endpoint = unsafe { CoapTcpEndpoint::new(self, addr)? }.into();
+        let endpoint = CoapEndpoint::new_tcp_endpoint(self, addr)?.into();
```
Comment (JKRhb) [inline libcoap/src/context.rs:248]:
```suggestion
        let endpoint = CoapEndpoint::new_tcp_endpoint(self, addr)?;
```

File: libcoap/src/context.rs (hunk 1 above)
Comment (petardimitrijevic) [reply libcoap/src/context.rs]: Fixed. No need for cast now that there is only one structure.

File: libcoap/src/context.rs (hunk 2 above)
Comment (petardimitrijevic) [reply libcoap/src/context.rs]: You are right. It is redundant. Removed.

**Comment by petardimitrijevic** [comment]
@JKRhb Thank you for your comments. While I was looking at them I figured out that endpoint creation can be simplified further. I've removed all `CoapEndpoint::new_endpoint_{udp,tcp,dtls}` methods and just left the `new_endpoint` method. This method is now public for the crate.

Further, I've modified the context to call the `CoapEndpoint::new_endpoint` method and pass the required protocol.
This further reduces the boilerplate code for endpoint creation. Further more it is now completely contained in `CoapContext::add_endpoint` method.

**Review by pulsastrix — APPROVED** [review]
Sorry for getting to this somewhat late, and thanks for the contribution.

Originally, I wanted to have separate structs for different endpoint types in case there are specific libcoap-functions that only apply to a subset of them, but as this doesn't seem to be the case simplifying the endpoints into a single struct makes sense.

LGTM

**Review by JKRhb — APPROVED** [review]

**Review by falko17 — APPROVED** [review]

**Pull request merged**